google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/rstudio/RStudioCredentialTester.java

```diff
@@ -0,0 +1,252 @@
+  }
+
+  @Override
+  public String description() {
+    return "RStudio credential tester.";
+  }
+
+  private static String buildTargetUrl(NetworkService networkService, String path) {
+    StringBuilder targetUrlBuilder = new StringBuilder();
+
+    if (NetworkServiceUtils.isWebService(networkService)) {
+      targetUrlBuilder.append(NetworkServiceUtils.buildWebApplicationRootUrl(networkService));
+    } else {
+      // Default to HTTP protocol when the scanner cannot identify the actual service.
+      targetUrlBuilder
+          .append("http://")
+          .append(NetworkEndpointUtils.toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    targetUrlBuilder.append(path);
+    return targetUrlBuilder.toString();
+  }
+
+  /**
+   * Determines if this tester can accept the {@link NetworkService} based on the name of the
+   * service or a custom fingerprint. The fingerprint is necessary since nmap doesn't recognize a
+   * rstudio server instance correctly.
+   *
+   * @param networkService the network service passed by tsunami
+   * @return true if a rstudio server instance is recognized
+   */
+  @Override
+  public boolean canAccept(NetworkService networkService) {
+    boolean canAcceptByNmapReport =
+        NetworkServiceUtils.getWebServiceName(networkService).equals(RSTUDIO_SERVICE);
+    if (canAcceptByNmapReport) {
+      return true;
+    }
+    boolean canAcceptByCustomFingerprint = false;
+    String url = buildTargetUrl(networkService, "unsupported_browser.htm");
+    try {
+      logger.atInfo().log("Probing RStudio - custom fingerprint phase");
+      HttpResponse response = httpClient.send(get(url).withEmptyHeaders().build());
+      canAcceptByCustomFingerprint =
+          response.status().isSuccess()
+              && response.headers().get(SERVER_HEADER).isPresent()
+              && response.headers().get(SERVER_HEADER).get().equals(RSTUDIO_HEADER)
+              && response
+                  .bodyString()
+                  .map(RStudioCredentialTester::bodyContainsRStudioElements)
+                  .orElse(false);
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Unable to query '%s'.", url);
+      return false;
+    }
+    return canAcceptByCustomFingerprint;
+  }
+
+  private static boolean bodyContainsRStudioElements(String responseBody) {
+    Document doc = Jsoup.parse(responseBody);
+    String title = doc.title();
+    String p =
+        doc.body().getElementsByTag("p").first().outerHtml().split("<p>")[1].split("</p>")[0];
+
+    if (title.contains(RSTUDIO_UNSUPPORTED_BROWSER_TITLE)
+        && p.contains(RSTUDIO_UNSUPPORTED_BROWSER_P)) {
+      logger.atInfo().log("Found RStudio endpoint");
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public ImmutableList<TestCredential> testValidCredentials(
+      NetworkService networkService, List<TestCredential> credentials) {
+
+    return credentials.stream()
```

Comment:
I noticed that this tester takes a long time to complete, can we make it return once it finds a valid credential? 

Similar to https://github.com/google/tsunami-security-scanner-plugins/blob/4ce380aea042ef0b75c273447728a81e414139b4/google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/jenkins/JenkinsCredentialTester.java#L77-L81

Reply:
The test takes a long time since the default configuration of RStudio uses RSA, which takes time to decrypt and verify. I've implemented the required change.